python/pyspark/sql/types.py

```diff
@@ -935,6 +936,86 @@ def _parse_datatype_json_value(json_value):
         long: LongType,
     })
 
+# Mapping Python array types to Spark SQL DataType
+# We should be careful here. The size of these types in python depends on C
+# implementation. We need to make sure that this conversion does not lose any
+# precision.
+#
+# Reference for C integer size, see:
+# ISO/IEC 9899:201x specification, § 5.2.4.2.1 Sizes of integer types <limits.h>.
+# Reference for python array typecode, see:
+# https://docs.python.org/2/library/array.html
+# https://docs.python.org/3.6/library/array.html
+
+_array_int_typecode_ctype_mappings = {
+    'b': ctypes.c_byte,
+    'B': ctypes.c_ubyte,
+    'h': ctypes.c_short,
+    'H': ctypes.c_ushort,
+    'i': ctypes.c_int,
+    'I': ctypes.c_uint,
+    'l': ctypes.c_long,
+    'L': ctypes.c_ulong
+}
+
+# TODO: Uncomment this when 'q' and 'Q' are supported by net.razorvine.pickle
+# Type code 'q' and 'Q' are not available at python 2
+# if sys.version > "2":
+#     _array_int_typecode_ctype_mappings.update({
+#         'q': ctypes.c_longlong,
+#         'Q': ctypes.c_ulonglong
+#     })
+
+
+def _int_size_to_type(size):
+    """
+    Return the Scala type from the size of integers.
+    """
+    if size <= 8:
+        return ByteType
+    if size <= 16:
+        return ShortType
+    if size <= 32:
+        return IntegerType
+    if size <= 64:
+        return LongType
+    raise TypeError("not supported type: integer size too large.")
```

Comment:
I don't think we should log this. This is just a helper function that helps to construct `_array_type_mappings`, which is a complete list of all supported type codes. Being filtered out here is not an error, it's by design. If users try to use unsupported typecode, they will see another `TypeError` due to line 1052:
```python
.....
elif isinstance(obj, array):
    if obj.typecode in _array_type_mappings:
        return ArrayType(_array_type_mappings[obj.typecode](), True)
    else:
        raise TypeError("not supported type: array(%s)" % obj.typecode)
.....
```